docs/v1/api/Config.md
<!-- title: Config; order: 15.2 -->

### Config

This interface includes all possible options that can be set in the
configuration file or via `app.config` entry.

NOTE: This interface is not exposed directly, it's augmented into the global
`app` namespace, so use `app.Config` instead, further more, you can augment any
number of options into this interface is you want.

```ts
interface app.Config { }
```

### lang

- \<string\> Default language of the application.

### saveSchedules

- \<boolean> Save schedules when the system shuts down and rebuild after reboot.

### statics

- \<string\> | \<{ [path: string]: [StaticOptions](#StaticOptions) }\>

where <code id="StaticOptions">StaticOptions</code> is:

```ts
interface StaticOptions extends serveStatic.ServeStaticOptions {
    /** 
     * If `true`, the URL must contain the folder name (relative to `SRC_PATH`) 
     * as prefix to reach the static resource. Also you can set a specified 
     * prefix other than the folder name.
     */
    prefix?: boolean | string;
    setHeaders?: (res: ServerResponse, path: string, stat: Stats) => void;
};
```

- [\<serveStatic.ServeStaticOptions\>](https://github.com/expressjs/serve-static#options)

### watch

Watch modules and when files changed, refresh the memory cache and hot-reload
the module.

NOTE: **DO NOT** import the module statically in anywhere, otherwise it may not
be reloaded as expected.

- <{ watch: (...args: any[]) => [chokidar.FSWatcher](https://github.com/paulmillr/chokidar#methods--events) }[]>

### server

Includes settings related to the web server or RPC servers.

#### server.hostname

- \<string\> | \<string[]\> Host name(s), used for calculating the sub-domain.

#### server.http

Includes the settings of the HTTP(s/2) server.

##### server.http.type

Sets the server type, possible values are:

- `http` (default)
- `https`
- `http2`

##### server.http.port

- \<number\> Sets the server port, default value is `80`.

##### server.http.timeout

- \<number\> Sets the default request timeout, default value is `120000` ms.

##### server.http.options

- \<object\> These options are mainly for type `http` and type `http2`.
    - [\<http.ServerOptions\>](https://nodejs.org/dist/latest-v15.x/docs/api/http.html#http_http_createserver_options_requestlistener)
    - [\<https.ServerOptions\>](https://nodejs.org/dist/latest-v15.x/docs/api/https.html#https_https_createserver_options_requestlistener)
    - [\<http2.ServerOptions\>](https://nodejs.org/dist/latest-v15.x/docs/api/http2.html#http2_http2_createsecureserver_options_onrequesthandler)

#### server.websocket

Includes the options of the WebSocket server.

##### server.websocket.enabled

- \<boolean\> Whether or not to turn on the WebSocket server.

##### server.websocket.port

- \<number\> By default, this `port` is `0` or `undefined`, that means it will
    attach to the HTTP server instead. If you change it, it will listen to that
    port instead.

##### server.websocket.options

- \<object\> These options are used for socket.io.
    - [\<SocketIO.ServerOptions\>](https://socket.io/docs/server-api/#new-Server-options)

#### server.rpc

- \<{ [appId: string]: object }\> Includes the settings of the RPC servers.
    - [\<microse.ChannelOptions\>](https://github.com/microse-rpc/microse-node/blob/master/docs/api.md#channeloptions)
    - [\<microse.ClientOptions\>](https://github.com/microse-rpc/microse-node/blob/master/docs/api.md#clientoptions)
    - `services` [\<ModuleProxy[]\>](https://github.com/microse-rpc/microse-node/blob/master/docs/api.md#moduleproxy)
        The services that should be hosted by this server.
    - `dependencies` <"all"> | [\<ModuleProxy[]\>](https://github.com/microse-rpc/microse-node/blob/master/docs/api.md#moduleproxy)
        The services that this server depended on. (hosted ones excluded.)


HISTORY.md
## 0.1.20

## Fixes

- Fix bugs in session handler.

## New Features

- Support dependency injection backed by [injectable-ts](https://github.com/hyurl/injectable-ts).
- Add two filter decorators `@before()` and `@after()` to add filter functions
    before/after the actual method is called.

## 0.1.14

## Fixes

- Fix a bug in `MarkdownParser` (and `HttpController.viewMarkdown()`).

# 0.1.13

## Fixes

- Fix the bug that after calling `res.end()`, `res.send()` or `res.redirect()`,
    the framework will still try to send data.

## New Features

- **DI (Dependency Injection)** support for controllers, now can set URL params
    as method parameter, and pass `req: Request`, `res: Response`, or 
    `socket: WebSocket` in arbitrary orders.
- Add two special methods `Controller.before()` and `Controller.after()` to 
    handle any operations before and after the actual method is called.

# 0.1.12

## Fixes

- Fix a bug in command-line execution.

# 0.1.11

## Fixes

- Fix the bug of `@requireAuth` not working with `WebSocketController`s.

## New Features

- Expose/load entities more reasonably, many variables, constants and 
    functions are now available in `config.ts`. 
- `DevWatcher` will delay restarting the server when more than one file are 
    changed at once.
- Support `.env` file (in `ROOT_PATH`).

# 0.1.3

## Changes

- `config.staticPath` => `config.statics` support multiple directories.
- `config.server.host` => `config.server.hostname`
- `config.server.port` => `config.server.http.port`
- `config.server.socket` => `config.server.websocket`

## New Features

- `config.awaitGenerator` Allows generator method in a controller being called
    as a coroutine function.

src/config.ts
import * as Session from "express-session";
import { ServerResponse } from "http";
import { Stats } from "fs";
import serveStatic = require("serve-static");
import { ServerOptions } from "socket.io";
import * as https from "https";
import * as http2 from "http2";
import { ChannelOptions, ClientOptions, ModuleProxy } from "microse";
import { FSWatcher } from "chokidar";

declare global {
    namespace app {
        const config: Config;
        interface Config {
            /** Default language of the application. */
            lang?: string;
            /**
             * Save schedules when the system shuts down and rebuild after
             * reboot.
             */
            saveSchedules?: boolean;
            /**
             * The directories that serve static resources.
             * @see https://www.npmjs.com/package/serve-static
             */
            statics?: string[] | { [path: string]: StaticOptions };
            /**
             * Watch modules and when files changed, refresh the memory cache
             * and hot-reload the module.
             * 
             * NOTE: **DO NOT** import the module statically in anywhere,
             * otherwise it may not be reloaded as expected.
             * 
             * @see https://github.com/microse-rpc/microse-node
             */
            watch?: { watch: (...args: any[]) => FSWatcher }[];
            server: {
                /** Host name(s), used for calculating the sub-domain. */
                hostname?: string | string[];
                /** Includes the settings of the HTTP(s/2) server. */
                http?: {
                    /** Server type, default value: `http`. */
                    type?: "http" | "https" | "http2";
                    /** Default value is `80`. */
                    port?: number;
                    /** HTTP request timeout, default value is `120000`. */
                    timeout?: number;
                    /**
                     * These options are mainly for type `http` and type `http2`.
                     * @see https://nodejs.org/dist/latest-v14.x/docs/api/https.html#https_https_createserver_options_requestlistener
                     * @see https://nodejs.org/dist/latest-v14.x/docs/api/http2.html#http2_http2_createserver_options_onrequesthandler
                     * @see https://nodejs.org/dist/latest-v14.x/docs/api/tls.html#tls_tls_createsecurecontext_options
                     */
                    options?: https.ServerOptions & http2.ServerOptions;
                };
                /** Includes the settings of WebSocket server. */
                websocket?: {
                    /** Whether or not to turn on the WebSocket server. */
                    enabled?: boolean;
                    /**
                     * By default, this `port` is `0` or `undefined`, that means
                     * it will attach to the HTTP server instead. If you change
                     * it, it will listen to that port instead.
                     */
                    port?: number;
                    /**
                     * Options for SocketIO.
                     * @see https://socket.io
                     */
                    options?: ServerOptions;
                };
                /**
                 * Includes the settings of the RPC servers.
                 * @see https://github.com/microse-rpc/microse-node
                 */
                rpc?: {
                    [appId: string]: ChannelOptions & ClientOptions & {
                        [x: string]: any;
                        /** The services that should be hosted by this server. */
                        services: ModuleProxy<any>[];
                        /**
                         * The services that this server depended on. (hosted 
                         * ones excluded.)
                         */
                        dependencies?: "all" | ModuleProxy<any>[];
                    }
                };
            };
            /**
             * Configurations for Express-Session.
             * @see https://www.npmjs.com/package/express-session
             */
            session?: Session.SessionOptions;
        }
    }
}

/**
 * @see https://www.npmjs.com/package/serve-static
 */
export interface StaticOptions extends serveStatic.ServeStaticOptions {
    /** 
     * If `true`, the URL must contain the folder name (relative to `SRC_PATH`) 
     * as prefix to reach the static resource. Also you can set a specified 
     * prefix other than the folder name.
     */
    prefix?: boolean | string;
    setHeaders?: (res: ServerResponse, path: string, stat: Stats) => void;
};

/**
 * The configuration of the program.
 * Some of these settings are for their dependencies, you may check out all 
 * supported options on their official websites.
 */
export default <app.Config>{
    lang: "en-US",
    saveSchedules: false,
    statics: ["assets"],
    watch: [],
    server: {
        hostname: "localhost",
        http: {
            type: "http",
            port: 4000,
            timeout: 120000, // 2 min.
            options: null
        },
        websocket: {
            enabled: true,
            port: undefined,
            options: {
                pingTimeout: 5000,
                pingInterval: 5000
            },
        },
        rpc: {}
    },
    session: null
};
